File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        left = 0
        right = len(nums) - 1


        while left <= right:

            current = (left + right) // 2

            if (nums[current] > nums[right]):
                left = current + 1
            elif(nums[current] < nums[right] and nums[current - 1] < nums[current]):
                right = current - 1
            else:
                return nums[current]    

File: Data Structures & Algorithms/trapping-rain-water/submission-1.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;

        int maxLeft = height[left];
        int maxRight = height[right];

        int water = 0;

        while(left < right) {
            
            if (maxLeft <= maxRight) {
                left++;
                maxLeft = Math.max(maxLeft, height[left]);
                water += Math.max(0, maxLeft - height[left]);
            } else {
                right--;
                maxRight = Math.max(maxRight, height[right]);
                water += Math.max(0, maxRight - height[right]);
            }

            
        }

        return water;

    }
}
